services/wallet/thirdparty/coingecko/client.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	url := fmt.Sprintf("%scoins/list", baseURL)
+	resp, err := c.DoQuery(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var tokens []GeckoToken
+	err = json.Unmarshal(body, &tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, token := range tokens {
+		if id, ok := coinGeckoMapping[strings.ToUpper(token.Symbol)]; ok {
+			if id != token.ID {
+				continue
+			}
+		}
+		c.tokens[strings.ToUpper(token.Symbol)] = token
+	}
+
+	return c.tokens, nil
+}
+
+func (c *Client) mapSymbolsToIds(symbols []string) ([]string, error) {
+	tokens, err := c.getTokens()
+	if err != nil {
+		return nil, err
+	}
+	ids := make([]string, 0)
+	for _, symbol := range utils.RenameSymbols(symbols) {
+		if token, ok := tokens[symbol]; ok {
+			ids = append(ids, token.ID)
+		}
+	}
+	return ids, nil
+}
+
+func (c *Client) getIDFromSymbol(symbol string) (string, error) {
+	tokens, err := c.getTokens()
+	if err != nil {
+		return "", err
+	}
+
+	return tokens[strings.ToUpper(symbol)].ID, nil
+}
+
+func (c *Client) FetchPrices(symbols []string, currencies []string) (map[string]map[string]float64, error) {
+	ids, err := c.mapSymbolsToIds(symbols)
+	if err != nil {
+		return nil, err
+	}
+	url := fmt.Sprintf("%ssimple/price?ids=%s&vs_currencies=%s", baseURL, strings.Join(ids, ","), strings.Join(currencies, ","))
+	resp, err := c.DoQuery(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	prices := make(map[string]map[string]float64)
+	err = json.Unmarshal(body, &prices)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]map[string]float64)
+	for _, symbol := range symbols {
+		result[symbol] = map[string]float64{}
+		id, err := c.getIDFromSymbol(utils.GetRealSymbol(symbol))
+		if err != nil {
+			return nil, err
+		}
+		for _, currency := range currencies {
+			result[symbol][currency] = prices[id][strings.ToLower(currency)]
+		}
+	}
+
+	return result, nil
+}
+
+func (c *Client) FetchTokenDetails(symbols []string) (map[string]thirdparty.TokenDetails, error) {
+	tokens, err := c.getTokens()
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]thirdparty.TokenDetails)
+	for _, symbol := range symbols {
+		if value, ok := tokens[utils.GetRealSymbol(symbol)]; ok {
+			result[symbol] = thirdparty.TokenDetails{
+				ID:     value.ID,
+				Name:   value.Name,
+				Symbol: symbol,
+			}
+		}
+
+	}
+	return result, nil
+}
+
+func (c *Client) FetchTokenMarketValues(symbols []string, currency string) (map[string]thirdparty.TokenMarketValues, error) {
+	ids, err := c.mapSymbolsToIds(symbols)
+	if err != nil {
+		return nil, err
+	}
+	url := fmt.Sprintf("%scoins/markets?ids=%s&vs_currency=%s&order=market_cap_desc&per_page=250&page=1&sparkline=false&price_change_percentage=1h,24h", baseURL, strings.Join(ids, ","), currency)
```

Comment:
This is just me being annoying so feel free to ignore the comment, but you can make url building more elegant with [url.Values](https://pkg.go.dev/net/url#Values):
```go
queryParams := url.Values{
    "id":  {string.Join(ids, ",")},
    "vs_currency": {currency},
    "order": {"market_cap_desc"}
     etc etc etc
}
url := "baseURL" + "coins/markets" + queryParams.Encode() 
```

Reply:
updated